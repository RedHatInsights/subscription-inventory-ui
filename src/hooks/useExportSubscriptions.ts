import { useQuery } from 'react-query';
import { HttpError } from '../utilities/errors';
import { useToken } from '../utilities/platformServices';

interface ExportSubscriptionsBody {
  blob: Blob;
  filename: string;
}

const parseFilename = (headers: any): string => {
  return headers.get('Content-Disposition')?.split(';')[1]?.split('=')[1]?.replaceAll('"', '');
};

const fetchExportSubscriptionsData = (jwtToken: Promise<string>) => async () => {
  const response = await fetch('/api/rhsm/v2/products/export', {
    headers: { Authorization: `Bearer ${await jwtToken}` }
  });

  if (!response.ok) {
    throw new HttpError('Failed to download export', response.status, response.statusText);
  }
  const blob = await response.blob();
  const filename = parseFilename(response.headers);

  return { blob, filename } as ExportSubscriptionsBody;
};

const useExportSubscriptions = () => {
  const jwtToken = useToken();
  return useQuery('export', () => fetchExportSubscriptionsData(jwtToken)(), {
    enabled: false,
    retry: 0
  });
};

export { useExportSubscriptions as default };
