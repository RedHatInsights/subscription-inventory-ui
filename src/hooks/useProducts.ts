import { useQuery, QueryObserverResult } from 'react-query';
import Cookies from 'js-cookie';

type Product = {
  name: string;
  productLine: string;
  quantity: number;
  sku: string;
};

interface ProductApiData {
  body: Product[];
}

const fetchProductData = async (): Promise<Product[]> => {
  const jwtToken = Cookies.get('cs_jwt');

  const response = await fetch('/api/rhsm/v2/products', {
    headers: { Authorization: `Bearer ${jwtToken}` }
  });

  const productResponseData: ProductApiData = await response.json();

  return productResponseData.body;
};

const getProducts = async (): Promise<Product[]> => {
  const productData = await fetchProductData();
  return productData;
};

const useProducts = (): QueryObserverResult<Product[], unknown> => {
  return useQuery('products', () => getProducts());
};

export { Product, useProducts as default };
