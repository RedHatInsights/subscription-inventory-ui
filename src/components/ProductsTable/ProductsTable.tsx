import React, { FunctionComponent } from 'react';
import { TableComposable, Thead, Tr, Th, Tbody, Td, ThProps } from '@patternfly/react-table';
import { v4 as uuid } from 'uuid';
import Unavailable from '@redhat-cloud-services/frontend-components/Unavailable';
import { Text, TextContent, TextVariants } from '@patternfly/react-core';
import { Processing } from '../../components/emptyState';
import useProducts, { Product } from '../../hooks/useProducts';

const ProductsTable: FunctionComponent = () => {
  const columnNames = { name: 'Name', quantity: 'Quantity' };
  const { isLoading, error, data } = useProducts();

  const [activeSortIndex, setActiveSortIndex] = React.useState<number | null>(null);

  const [activeSortDirection, setActiveSortDirection] = React.useState<'asc' | 'desc' | null>(null);

  const getSortableRowValues = (product: Product): (string | number)[] => {
    const { name, quantity } = product;
    return [name, quantity];
  };

  const getSortParams = (columnIndex: number): ThProps['sort'] => ({
    sortBy: {
      index: activeSortIndex,
      direction: activeSortDirection,
      defaultDirection: 'asc' // starting sort direction when first sorting a column. Defaults to 'asc'
    },
    onSort: (_event, index, direction) => {
      setActiveSortIndex(index);
      setActiveSortDirection(direction);
    },
    columnIndex
  });

  const sortProducts = (products: Product[], sortIndex: number) => {
    if (sortIndex == null) return products;

    const sortedProducts = products.sort((a: any, b: any) => {
      const aValue = getSortableRowValues(a)[sortIndex];
      const bValue = getSortableRowValues(b)[sortIndex];
      if (typeof aValue === 'number') {
        // Numeric sort
        if (activeSortDirection === 'asc') {
          return (aValue as number) - (bValue as number);
        }
        return (bValue as number) - (aValue as number);
      } else {
        // String sort
        if (activeSortDirection === 'asc') {
          return (aValue as string).localeCompare(bValue as string);
        }
        return (bValue as string).localeCompare(aValue as string);
      }
    });
    return sortedProducts;
  };

  const Results: FunctionComponent = () => {
    const sortedProducts = sortProducts(data, activeSortIndex);

    return (
      <TableComposable aria-label="Products" ouiaId={uuid()} ouiaSafe={true}>
        <Thead>
          <Tr ouiaId={uuid()} ouiaSafe={true}>
            <Th sort={getSortParams(0)}>{columnNames.name}</Th>
            <Th sort={getSortParams(1)}>{columnNames.quantity}</Th>
          </Tr>
        </Thead>
        <Tbody>
          {sortedProducts.map((datum, rowIndex) => (
            <Tr key={rowIndex} ouiaId={uuid()} ouiaSafe={true}>
              <Td dataLabel={columnNames.name}>
                <TextContent>
                  <Text component={TextVariants.h3}>
                    {datum.name}
                    <br />
                    <Text component={TextVariants.small}>{datum.productLine}</Text>
                  </Text>
                </TextContent>
              </Td>
              <Td dataLabel={columnNames.quantity}>{datum.quantity}</Td>
            </Tr>
          ))}
        </Tbody>
      </TableComposable>
    );
  };

  if (isLoading && !error) {
    return <Processing />;
  } else if (!isLoading && !error) {
    return <Results />;
  } else {
    return <Unavailable />;
  }
};

export default ProductsTable;
