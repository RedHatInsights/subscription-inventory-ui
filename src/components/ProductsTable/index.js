import ProductsTable from './ProductsTable';

export { ProductsTable as default };
