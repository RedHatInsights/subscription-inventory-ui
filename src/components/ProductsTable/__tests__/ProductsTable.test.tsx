import React, { FunctionComponent } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import ProductsTable, { ProductsTableProps } from '../ProductsTable';
import { get, def } from 'bdd-lazy-var';
import factories from '../../../utilities/factories';
import { Product } from '../../../hooks/useProducts';

const queryClient = new QueryClient();

const Table: FunctionComponent<ProductsTableProps> = ({ data, isFetching, filter, setFilter }) => (
  <QueryClientProvider client={queryClient}>
    <BrowserRouter>
      <ProductsTable data={data} isFetching={isFetching} filter={filter} setFilter={setFilter} />
    </BrowserRouter>
  </QueryClientProvider>
);

const setFilter = jest.fn();

describe('ProductsTable', () => {
  def('loading', () => false);
  def('error', () => false);
  def('fetching', () => false);
  def('data', () => [
    factories.product.build({
      name: 'A',
      productLine: 'letters',
      sku: 'MOCK123',
      quantity: 3,
      serviceLevel: 'Standard'
    })
  ]);

  it('renders correctly', () => {
    const { getAllByText } = render(
      <Table
        data={get('data')}
        isFetching={get('fetching')}
        filter={get('filter')}
        setFilter={setFilter}
      />
    );

    getAllByText('A').forEach((el) => {
      expect(el).toBeInTheDocument();
    });
  });

  it('renders the export button', () => {
    const { container } = render(
      <Table
        data={get('data')}
        isFetching={get('fetching')}
        filter={get('filter')}
        setFilter={setFilter}
      />
    );
    const buttonEl = container.querySelector('#export-button');
    expect(buttonEl).not.toBeNull();
    expect(buttonEl).toBeInTheDocument();
  });

  describe('when row column headings are clicked', () => {
    const formattedData = [
      {
        name: 'Z',
        sku: 'RH1234',
        quantity: 1,
        productLine: 'letters',
        serviceLevel: 'Layered'
      },
      {
        name: 'A',
        sku: 'MOCK123',
        quantity: 3,
        productLine: 'vowels',
        serviceLevel: 'Standard'
      },
      {
        name: 'A',
        sku: 'MOCK123',
        quantity: 0,
        productLine: 'vowels',
        serviceLevel: 'Standard'
      },
      {
        name: '',
        sku: 'AMOCK1234',
        quantity: 2,
        productLine: 'consonants',
        serviceLevel: 'Standard'
      }
    ];

    it('sorts by name by default', () => {
      render(
        <Table
          data={formattedData as Product[]}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      expect(table.children[1].firstChild.childNodes[0].textContent).toEqual(
        formattedData[0].productLine
      );
    });

    it('can sort by name, reversed', () => {
      render(
        <Table
          data={formattedData as Product[]}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      fireEvent.click(screen.getByText('Name'));
      expect(table.children[1].firstChild.childNodes[0].textContent).toEqual(
        formattedData[0].name + formattedData[0].productLine
      );
    });

    it('can sort by quantity', () => {
      render(
        <Table
          data={formattedData as Product[]}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      fireEvent.click(screen.getByText('Quantity'));
      expect(parseInt(table.children[1].firstChild.childNodes[2].textContent)).toEqual(0);
    });

    it('can sort by quantity, reversed', () => {
      render(
        <Table
          data={formattedData as Product[]}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      fireEvent.click(screen.getByText('Quantity'));
      fireEvent.click(screen.getByText('Quantity'));
      expect(parseInt(table.children[1].firstChild.childNodes[2].textContent)).toEqual(3);
    });

    it('can sort by sku', () => {
      render(
        <Table
          data={formattedData as Product[]}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      fireEvent.click(screen.getByText('SKU'));
      expect(table.children[1].firstChild.childNodes[1].textContent).toEqual('AMOCK1234');
    });

    it('can sort by sku, reversed', () => {
      render(
        <Table
          data={formattedData as Product[]}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      fireEvent.click(screen.getByText('SKU'));
      fireEvent.click(screen.getByText('SKU'));
      expect(table.children[1].firstChild.childNodes[1].textContent).toEqual('RH1234');
    });

    it('can sort by serviceLevel', () => {
      render(
        <Table
          data={formattedData as Product[]}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      fireEvent.click(screen.getByText('Service level'));
      expect(table.children[1].firstChild.childNodes[3].textContent).toEqual('Layered');
    });

    it('can sort by serviceLevel, reversed', () => {
      render(
        <Table
          data={formattedData as Product[]}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      fireEvent.click(screen.getByText('Service level'));
      fireEvent.click(screen.getByText('Service level'));
      expect(table.children[1].firstChild.childNodes[3].textContent).toEqual('Standard');
    });
  });

  describe('when using pagination', () => {
    def('data', () => [
      ...factories.product.buildList(10, {
        name: 'A',
        sku: 'MOCK123',
        quantity: 1,
        productLine: 'letters',
        serviceLevel: 'Standard'
      }),
      factories.product.build({
        name: 'Z',
        sku: 'RH123',
        quantity: 2,
        productLine: 'letters',
        serviceLevel: 'Standard'
      })
    ]);

    it('can change page', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const nextPage = screen.getAllByLabelText('Go to next page')[0];
      fireEvent.click(nextPage);
      const currentPage = screen.getAllByLabelText('Current page');
      expect(currentPage[0]).toHaveValue(2);
    });

    it('can change per page', () => {
      const { container } = render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const table = screen.getByLabelText('Products');

      const perPageArrow = container.querySelectorAll('#options-menu-top-toggle')[0];
      fireEvent.click(perPageArrow);
      const perPageAmount = screen.getByText('20 per page');
      fireEvent.click(perPageAmount);
      expect(table.querySelectorAll('tr').length).toBe(12);
    });
  });

  describe('when the search is used', () => {
    def('data', () => [
      factories.product.build({
        name: 'Z',
        sku: 'RH123',
        quantity: 2,
        productLine: 'consonants',
        serviceLevel: 'Standard'
      }),
      factories.product.build({
        name: 'A',
        sku: 'MOCK123',
        quantity: 1,
        productLine: 'vowels',
        serviceLevel: 'Standard'
      }),
      factories.product.build({
        name: 'P',
        sku: 'Fake123',
        quantity: 1,
        productLine: 'consonants',
        serviceLevel: 'Standard'
      }),
      factories.product.build({
        name: undefined,
        sku: undefined,
        quantity: 0,
        productLine: undefined,
        serviceLevel: undefined
      })
    ]);

    it('refines the results by name', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const input = screen.getByPlaceholderText('Filter by Name or SKU');
      fireEvent.change(input, { target: { value: 'Z' } });
      expect(screen.queryByText('A')).not.toBeInTheDocument();
    });

    it('refines the results by product line', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const input = screen.getByPlaceholderText('Filter by Name or SKU');
      fireEvent.change(input, { target: { value: 'vowels' } });
      expect(screen.queryByText('consonants')).not.toBeInTheDocument();
    });

    it('can be cleared', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const input = screen.getByPlaceholderText('Filter by Name or SKU');
      fireEvent.change(input, { target: { value: 'Z' } });
      const clear = screen.getByLabelText('Reset');
      fireEvent.click(clear);
      expect(screen.queryByText('A')).toBeInTheDocument();
    });

    it('renders an empty state when no results are found', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const input = screen.getByPlaceholderText('Filter by Name or SKU');
      fireEvent.change(input, { target: { value: 'ZZZ' } });
      expect(screen.getByText('No results found')).toBeInTheDocument();
    });

    it('refines the results by sku', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={setFilter}
        />
      );

      const input = screen.getByPlaceholderText('Filter by Name or SKU');
      fireEvent.change(input, { target: { value: 'RH123' } });
      expect(screen.queryByText('Fake123')).not.toBeInTheDocument();
    });
  });

  describe('when the active filter is set', () => {
    def('data', () => [
      factories.product.build({
        name: 'Z',
        sku: 'RH123',
        quantity: 2,
        productLine: 'consonants',
        serviceLevel: 'Standard'
      }),
      factories.product.build({
        name: 'A',
        sku: 'MOCK123',
        quantity: 1,
        productLine: 'vowels',
        serviceLevel: 'Standard'
      })
    ]);

    it('renders with filter enabled', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={'active'}
          setFilter={() => {
            'active';
          }}
        />
      );

      expect(screen.getByText('Active')).toBeInTheDocument();
    });

    it('renders status filter', async () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={'active'}
          setFilter={setFilter}
        />
      );

      expect(screen.getByRole('button', { name: 'close Active' }));
    });

    it('renders Clear filers', async () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={'active'}
          setFilter={setFilter}
        />
      );
      expect(screen.getByRole('button', { name: 'Clear filters' })).toBeInTheDocument();
    });
  });

  describe('when the expiringSoon filter is set', () => {
    def('filter', () => 'expiringSoon');

    it('renders with filter enabled', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={() => {
            'expiringSoon';
          }}
        />
      );

      expect(screen.getByText('Expiring soon')).toBeInTheDocument();
    });
  });

  describe('when the expired filter is set', () => {
    def('filter', () => 'expired');

    it('renders with filter enabled', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={() => {
            'expired';
          }}
        />
      );

      expect(screen.getByText('Expired')).toBeInTheDocument();
    });
  });

  describe('when the futureDated filter is set', () => {
    def('filter', () => 'futureDated');

    it('renders with filter enabled', () => {
      render(
        <Table
          data={get('data')}
          isFetching={get('fetching')}
          filter={get('filter')}
          setFilter={() => {
            'futureDated';
          }}
        />
      );

      expect(screen.getByText('Future dated')).toBeInTheDocument();
    });
  });
});
