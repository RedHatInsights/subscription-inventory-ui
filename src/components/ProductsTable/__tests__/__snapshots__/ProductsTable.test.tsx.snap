// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsTable renders correctly 1`] = `
<div>
  <table
    aria-label="Products"
    class="pf-c-table pf-m-grid-md"
    data-ouia-component-id="00000000-0000-0000-0000-000000000000"
    data-ouia-component-type="PF4/Table"
    data-ouia-safe="true"
    role="grid"
  >
    <thead
      class=""
    >
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <th
          aria-sort="none"
          class="pf-c-table__sort"
          scope="col"
        >
          <button
            class="pf-c-table__button"
            type="button"
          >
            <div
              class="pf-c-table__button-content"
            >
              <span
                class="pf-c-table__text"
              >
                Name
              </span>
              <span
                class="pf-c-table__sort-indicator"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
        <th
          aria-sort="none"
          class="pf-c-table__sort"
          scope="col"
        >
          <button
            class="pf-c-table__button"
            type="button"
          >
            <div
              class="pf-c-table__button-content"
            >
              <span
                class="pf-c-table__text"
              >
                Quantity
              </span>
              <span
                class="pf-c-table__sort-indicator"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
      </tr>
    </thead>
    <tbody
      class=""
      role="rowgroup"
    >
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <td
          class=""
          data-label="Name"
        >
          <div
            class="pf-c-content"
          >
            <h3
              class=""
              data-ouia-component-id="OUIA-Generated-Text-1"
              data-ouia-component-type="PF4/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              A
              <br />
              <small
                class=""
                data-ouia-component-id="OUIA-Generated-Text-2"
                data-ouia-component-type="PF4/Text"
                data-ouia-safe="true"
                data-pf-content="true"
              >
                letters
              </small>
            </h3>
          </div>
        </td>
        <td
          class=""
          data-label="Quantity"
        >
          3
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ProductsTable when loading renders the loading state 1`] = `
<div>
  <div
    class="pf-l-bullseye"
  >
    <span
      aria-valuetext="Loading..."
      class="pf-c-spinner pf-m-xl"
      role="progressbar"
    >
      <span
        class="pf-c-spinner__clipper"
      />
      <span
        class="pf-c-spinner__lead-ball"
      />
      <span
        class="pf-c-spinner__tail-ball"
      />
    </span>
  </div>
</div>
`;

exports[`ProductsTable when row column headings are clicked can sort by name 1`] = `
<div>
  <table
    aria-label="Products"
    class="pf-c-table pf-m-grid-md"
    data-ouia-component-id="00000000-0000-0000-0000-000000000000"
    data-ouia-component-type="PF4/Table"
    data-ouia-safe="true"
    role="grid"
  >
    <thead
      class=""
    >
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <th
          aria-sort="ascending"
          class="pf-c-table__sort pf-m-selected"
          scope="col"
        >
          <button
            class="pf-c-table__button"
            type="button"
          >
            <div
              class="pf-c-table__button-content"
            >
              <span
                class="pf-c-table__text"
              >
                Name
              </span>
              <span
                class="pf-c-table__sort-indicator"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M88 166.059V468c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12V166.059h46.059c21.382 0 32.09-25.851 16.971-40.971l-86.059-86.059c-9.373-9.373-24.569-9.373-33.941 0l-86.059 86.059c-15.119 15.119-4.411 40.971 16.971 40.971H88z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
        <th
          aria-sort="none"
          class="pf-c-table__sort"
          scope="col"
        >
          <button
            class="pf-c-table__button"
            type="button"
          >
            <div
              class="pf-c-table__button-content"
            >
              <span
                class="pf-c-table__text"
              >
                Quantity
              </span>
              <span
                class="pf-c-table__sort-indicator"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
      </tr>
    </thead>
    <tbody
      class=""
      role="rowgroup"
    >
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <td
          class=""
          data-label="Name"
        >
          <div
            class="pf-c-content"
          >
            <h3
              class=""
              data-ouia-component-id="OUIA-Generated-Text-9"
              data-ouia-component-type="PF4/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              A
              <br />
              <small
                class=""
                data-ouia-component-id="OUIA-Generated-Text-10"
                data-ouia-component-type="PF4/Text"
                data-ouia-safe="true"
                data-pf-content="true"
              >
                BMW Volt
              </small>
            </h3>
          </div>
        </td>
        <td
          class=""
          data-label="Quantity"
        >
          3
        </td>
      </tr>
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <td
          class=""
          data-label="Name"
        >
          <div
            class="pf-c-content"
          >
            <h3
              class=""
              data-ouia-component-id="OUIA-Generated-Text-11"
              data-ouia-component-type="PF4/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              B
              <br />
              <small
                class=""
                data-ouia-component-id="OUIA-Generated-Text-12"
                data-ouia-component-type="PF4/Text"
                data-ouia-safe="true"
                data-pf-content="true"
              >
                Audi PT Cruiser
              </small>
            </h3>
          </div>
        </td>
        <td
          class=""
          data-label="Quantity"
        >
          2
        </td>
      </tr>
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <td
          class=""
          data-label="Name"
        >
          <div
            class="pf-c-content"
          >
            <h3
              class=""
              data-ouia-component-id="OUIA-Generated-Text-13"
              data-ouia-component-type="PF4/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              Z
              <br />
              <small
                class=""
                data-ouia-component-id="OUIA-Generated-Text-14"
                data-ouia-component-type="PF4/Text"
                data-ouia-safe="true"
                data-pf-content="true"
              >
                Bentley Jetta
              </small>
            </h3>
          </div>
        </td>
        <td
          class=""
          data-label="Quantity"
        >
          1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ProductsTable when row column headings are clicked can sort by quantity 1`] = `
<div>
  <table
    aria-label="Products"
    class="pf-c-table pf-m-grid-md"
    data-ouia-component-id="00000000-0000-0000-0000-000000000000"
    data-ouia-component-type="PF4/Table"
    data-ouia-safe="true"
    role="grid"
  >
    <thead
      class=""
    >
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <th
          aria-sort="none"
          class="pf-c-table__sort"
          scope="col"
        >
          <button
            class="pf-c-table__button"
            type="button"
          >
            <div
              class="pf-c-table__button-content"
            >
              <span
                class="pf-c-table__text"
              >
                Name
              </span>
              <span
                class="pf-c-table__sort-indicator"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
        <th
          aria-sort="ascending"
          class="pf-c-table__sort pf-m-selected"
          scope="col"
        >
          <button
            class="pf-c-table__button"
            type="button"
          >
            <div
              class="pf-c-table__button-content"
            >
              <span
                class="pf-c-table__text"
              >
                Quantity
              </span>
              <span
                class="pf-c-table__sort-indicator"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M88 166.059V468c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12V166.059h46.059c21.382 0 32.09-25.851 16.971-40.971l-86.059-86.059c-9.373-9.373-24.569-9.373-33.941 0l-86.059 86.059c-15.119 15.119-4.411 40.971 16.971 40.971H88z"
                  />
                </svg>
              </span>
            </div>
          </button>
        </th>
      </tr>
    </thead>
    <tbody
      class=""
      role="rowgroup"
    >
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <td
          class=""
          data-label="Name"
        >
          <div
            class="pf-c-content"
          >
            <h3
              class=""
              data-ouia-component-id="OUIA-Generated-Text-21"
              data-ouia-component-type="PF4/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              Z
              <br />
              <small
                class=""
                data-ouia-component-id="OUIA-Generated-Text-22"
                data-ouia-component-type="PF4/Text"
                data-ouia-safe="true"
                data-pf-content="true"
              >
                Chevrolet Fortwo
              </small>
            </h3>
          </div>
        </td>
        <td
          class=""
          data-label="Quantity"
        >
          1
        </td>
      </tr>
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <td
          class=""
          data-label="Name"
        >
          <div
            class="pf-c-content"
          >
            <h3
              class=""
              data-ouia-component-id="OUIA-Generated-Text-23"
              data-ouia-component-type="PF4/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              B
              <br />
              <small
                class=""
                data-ouia-component-id="OUIA-Generated-Text-24"
                data-ouia-component-type="PF4/Text"
                data-ouia-safe="true"
                data-pf-content="true"
              >
                Mercedes Benz 1
              </small>
            </h3>
          </div>
        </td>
        <td
          class=""
          data-label="Quantity"
        >
          2
        </td>
      </tr>
      <tr
        class=""
        data-ouia-component-id="00000000-0000-0000-0000-000000000000"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe="true"
      >
        <td
          class=""
          data-label="Name"
        >
          <div
            class="pf-c-content"
          >
            <h3
              class=""
              data-ouia-component-id="OUIA-Generated-Text-25"
              data-ouia-component-type="PF4/Text"
              data-ouia-safe="true"
              data-pf-content="true"
            >
              A
              <br />
              <small
                class=""
                data-ouia-component-id="OUIA-Generated-Text-26"
                data-ouia-component-type="PF4/Text"
                data-ouia-safe="true"
                data-pf-content="true"
              >
                Polestar Expedition
              </small>
            </h3>
          </div>
        </td>
        <td
          class=""
          data-label="Quantity"
        >
          3
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ProductsTable when there is an error renders the error state 1`] = `
<div>
  <div
    class="pf-c-empty-state pf-m-lg ins-c-empty-state__unavailable pf-m-redhat-font"
  >
    <div
      class="pf-c-empty-state__content"
    >
      <svg
        aria-hidden="true"
        class="pf-c-empty-state__icon"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 512 512"
        width="1em"
      >
        <path
          d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
        />
      </svg>
      <h5
        class="pf-c-title pf-m-lg"
        data-ouia-component-id="OUIA-Generated-Title-1"
        data-ouia-component-type="PF4/Title"
        data-ouia-safe="true"
      >
        This page is temporarily unavailable
      </h5>
      <div
        class="pf-c-empty-state__body"
      >
        Try refreshing the page. If the problem persists, contact your organization administrator or visit our
        <a
          href="https://status.redhat.com/"
          rel="noopener noreferrer"
          target="_blank"
        >
           status page
        </a>
         for known outages.
      </div>
    </div>
  </div>
</div>
`;
